File: Day 40 DSA (235)/Code.cpp
class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        int small = min(p->val, q->val);
        int large = max(p->val, q->val);
        while (root != nullptr) {
            if (root->val > large) 
            //  left subtree
                root = root->left;
            else if (root->val < small) 
            // right subtree
                root = root->right;
            else 
            //small <= root.val <= large -> This root is the LCA between p and q
                return root;
        }
        return nullptr;
    }
};

File: Day30 DSA (543)/Code.py
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        def diameter(node, res):
            if not node:
                return 0
            left = diameter(node.left, res)
            right = diameter(node.right, res)
            res[0] = max(res[0], left + right)
            return max(left, right) + 1
        res = [0]
        diameter(root, res)
        return res[0]
